🚧 Chunked uploads handling (#80)

pvanliefland opened the pull request:
This PR addresses #80 

This implementation uses `contextvars` to store the chunks in-memory until the upload of the last chunk (at this point we send the file to S3 or GCS).

**Note:** using contextvars requires Python 3.7.

It should work for both S3 and GCS.

Todo:

- [x] Add `chunks` utility module
- [x] Adapt `save()` in generic manager
- [x] Periodic cleanup of the chunks stored in-memory
- [x] Tests
- [x] Import `contextvars` in try/except and disable chunks for Python < 3.7

pvanliefland commented:
Hey @ericdill, obviously a work in progress, but wanted to create the PR so you can have a look.

Couldn't find a way to run the tests on my laptop so the PR will help with that too.

pvanliefland commented on s3contents/ipycompat.py:
```diff
@@ -17,6 +17,7 @@
 from notebook.services.contents.filemanager import FileContentsManager
 from notebook.services.contents.manager import ContentsManager
 from notebook.services.contents.tests.test_contents_api import APITest
+from notebook.services.contents.tests.test_largefilemanager import TestLargeFileManager
```
@ericdill this is failing right now - can you think of anything that might prevent this import from working? Feel like I'm missing something... It seems to be here: https://github.com/jupyter/notebook/tree/master/notebook/services/contents/tests

ericdill replied:
version 4.* of the notebook didn't have this import. It was introduced in 5.0. This library tests for notebook 4.*, which I think it's time to drop. cc @danielfrg

danielfrg replied:
No problem dropping notebook 4.*

ericdill commented:
Couple of things that are failing. 
1. Tests should be updated to be only 3.7 or newer.
1. Notebook dependency 4.* should be dropped. This is where your ImportError is coming from

pvanliefland commented:
Ok, so a bit ashamed but as I couldn't find a way to run the tests locally, I use the Github actions to make sure that the tests pass.

I will of course rebase those ugly commits.

I think it is ready for another review @ericdill - I have added some tests and a basic check for staled chunked uploads.

The lint fails, I don't understand the issue with isort, on my local machine it does not generate any error...

In the meantime I will test it using a real Jupyter environment.

ericdill commented:
Ok I'll have a look later today.

To run tests locally, I think you need to do the following:

1. `make env`
2. `conda activate s3contents`
3. In one terminal, run `make minio` (after conda activate s3contents)
4. In another, run `make test-all` (after conda activate s3contents

Yeah I'm not sure what's up with the lint. Master branch has an error that I can't reproduce locally. *Might* be a different python version locally vs GH actions. I'll poke at that later today.

pvanliefland commented:
Ok, tested the fork in a "real" project and it seems to work

danielfrg commented on .github/workflows/test.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
       matrix:
         python-version: [3.6, 3.7, 3.8]
-        jupyter-version: ["4.*", "5.6", "5.7", "5.*", "6.0", "6.*"]
+        jupyter-version: ["5.6", "5.7", "5.*", "6.0", "6.*"]
```
Can you set the `requirements-package.txt` to `notebook>=5`

ericdill replied:
should we do `notebook>=5.6` since that's what's being tested?

pvanliefland replied:
I changed it to `notebook>=5.6`

ericdill commented on .github/workflows/test.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
         python-version: [3.6, 3.7, 3.8]
```
Don't we need to drop 3.6 too?

danielfrg replied:
i guess thats ok.

You can also just comment out the github actions step for the linting. Not a big issue.

If you change the python and other deps could you reflect those changes in setup.py and requirements-package.txt?

pvanliefland replied:
Regarding Python 3.6, I have tried to do it in a BC-way: if `contextvars` cannot be imported, the "chunked upload" feature is not available.

Of course I'm ok with removing compatibility with Python 3.6 altogether but it seemed a bit harsh.

Let me know what you decide.

pvanliefland replied:
Commented out Black in Makefile "check" for now, as it is causing the lint issue.

Locally, black tells me that everything is fine...

ericdill replied:
Can you comment out lines 62-64 in the github actions file instead? https://www.github.com/danielfrg/s3contents/tree/master/.github%2Fworkflows%2Ftest.yml

pvanliefland replied:
Done

ericdill commented:
Hi @pvanliefland @danielfrg,

Regarding backwards compat for python 3.6 specifically, I'm torn. On the one hand, supporting more python versions would be good. On the other, silent failures on file uploads is bad.

Is silent failures on uploads bad enough that we want to drop python 3.6 compatibility? Or should we maintain python 3.6 compat with the understanding that file uploads is broken in py3.6?

ericdill commented on .github/workflows/test.yml:
```diff
@@ -59,9 +59,9 @@ jobs:
-    - name: Check linting
-      run: |
-        make check
+    # - name: Check linting
```
Could you add a note here that we're commenting until #102 is resolved?

pvanliefland replied:
Done

ericdill commented:
Once this PR is in, you might consider submitting a PR to the [chunked-saving](https://jupyter-notebook.readthedocs.io/en/stable/extending/contents.html#chunked-saving) part of the notebook docs that references this as an example implementation of chunked saving.

danielfrg commented:
> Is silent failures on uploads bad enough that we want to drop python 3.6 compatibility?

I am also torn but I think that is bad enough, at least on my book. I am also in general happy to just drop things without thinking much about it lol, but I can be convinced otherwise

ericdill commented:
@danielfrg is _might_ not be a silent failure? Looks like this code path will 500. I wonder if that's seen by the user.

```python
        if not CHUNKS_HANDLING:
            self.log.error(
                "S3contents.GenericManager._save_large_file: not available with Python < 3.7"
            )
            self.do_error("Chunked file upload in S3/GCS requires Python >= 3.7", 500)
```

danielfrg commented:
Ah, i see. If we log it and show it to the notebook UI then I think we should keep Python 3.6.

pvanliefland commented:
@ericdill @danielfrg unfortunately, I've tested how errors are handled and it seems that although they appear in the console, they are not shown displayed to the user.

It means that in the context of chunked uploads, the file won't be uploaded at all (compared to now, where a truncated version of the file is uploaded), no message will be displayed, but the browser console prints a meaningful message.

Let me know how you want to proceed.

danielfrg commented:
Thanks for all the work on this!

In that case my vote would be to drop python 3.6. If you agree @ericdill then we can just do that.

ericdill commented:
yea, agreed. let's say farewell to python 3.6

ericdill commented:
I just tried to use your branch and things worked as expected 🎉  when the folder in the contents manager was a path to an s3 bucket. When I tried to upload to a directory backed by a FileContentsManager, the same behavior happens (truncated uploads). I suspect we should add some guidance to the README here in this repo that you should consider using the `LargeFileManager` instead of the `FileContentsManager` if you're having trouble with uploads.

pvanliefland commented:
@ericdill @danielfrg Ok, this should be ready to merged.

Latest changes:

- Dropped Python 3.6 compat
- Updated README so that it uses `LargeFileManager` instead of `FileContentsManager` (`LargeFileManager` has been the default manager in Jupyter notebook for a while now)
- Rebased all commits into one

ericdill reviewed (approved):
LGTM, @danielfrg?

ericdill commented:
Thanks for the pr @pvanliefland ! Going to assume @danielfrg is ok with this :) We can always fix any issues @danielfrg has in a follow-on PR.

pvanliefland commented:
Cool!
